### PR TITLE
[export] handle new roots & root swapping in derived dims suggested fixes

### PR DESCRIPTION
**Pull request by pianpwk**
Pull Request resolved: https://github.com/pytorch/pytorch/pull/125543

This PR address 2 issues with derived dim suggested fixes, 1) newly introduced roots, and 2) root swapping.

1 | Newly introduced roots appear with modulo guards, e.g. Mod(dx, 2) = 0 suggests dx is a derived dim equal to 2 * _dx, introducing a new root _dx. Currently the final suggested fixes handle this correctly, but we can get intermediate results where related derived dims don't rely on a unified root, and are a mixture of min/max range and derived suggestions. 

For example:
```
"dx": {"eq": 3*_dx-1, "max": 36}
"dy": {"eq": dx+1}
This should lead to suggested fixes
  _dx = Dim('_dx', max=12)
  dx = 3 * _dx - 1
  dy = 3 * _dx
```

This PR prettifies the suggested fixes routine by unifying to a single root, and making each intermediate suggestion either a derived dim or min/max range, not both.

2 | The current suggested fixes for derived dims can lead to root dims/derived dims being swapped, e.g. `dy - 1, dy` -> `dx, dx + 1`. This leads to problematic suggested fixes that look like `dy - 1 = Dim("dy - 1")` since we don't have access to the original variable name.

This PR only adds a suggested fix for the root dim, and removes all other derived suggestions.

For example, with the export test case test_derived_dim_out_of_order_simplified:
```
_dimz = torch.export.Dim("_dimz", min=6, max=8)
dimy = _dimz - 1
dimx = dimy - 1
dimz = torch.export.Dim("dimz", min=6, max=8)  # doesn't work, should be = _dimz

class Foo(torch.nn.Module):
    def forward(self, x, y, z):
        return x + y[1:] + z[2:]

foo = Foo()
u, v, w = torch.randn(5), torch.randn(6), torch.randn(7)
export(
    foo,
    (u, v, w),
    dynamic_shapes=({0: dimx}, {0: dimy}, {0: dimz}),
)
```

Before:
```
Suggested fixes:
  _dimz = Dim('_dimz', min=3, max=9223372036854775807)  # 2 <= _dimz - 1 <= 9223372036854775806
  _dimz - 2 = Dim('_dimz - 2', min=4, max=6)
  _dimz = Dim('_dimz', min=2, max=9223372036854775806)  # 2 <= _dimz <= 9223372036854775806
  _dimz - 1 = _dimz - 1
  dimz = _dimz
```

New suggested fixes:
```
Suggested fixes:
  dimz = _dimz
```

Note: This assumes the specified derived relations between dims are correct. This should be valid because: 1) if the relation is plain wrong (e.g. (dx, dx - 1) provided with inputs (6, 4)), this gets caught in beforehand in produce_guards. 2) if the relation is correct but does not match the emitted guard, for example:
```
def forward(self, x, y):
    return x.reshape([-1]) + y  # guard: s0 * 2 = s1
dx = Dim("dx")
export(
    model,
    (torch.randn(6, 2), torch.randn(12)),
    dynamic_shapes={"x": (dx, 2), "y": (dx + 6, )}
)
```
This produces two linear equations, leading to specialization since a) produce_guards is able to solve for a concrete value, and b) the export constraint solver will anyways force specializations due to range constraints.
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1334,6 +1334,16 @@ def cast_symbool_to_symint_guardless(symbool: torch.SymBool) -> torch.SymInt:
     int_sym = _sympy_cast_symbool_to_symint_guardless(symbool.node.expr)
     return symbool.node.shape_env.create_symintnode(int_sym, hint=int(symbool.node.require_hint()) if has_hint(symbool) else None)
 
+# use class name to avoid circular imports
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:1337]: eh, couldn't you do a local import to avoid circular imports?

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1334,6 +1334,16 @@ def cast_symbool_to_symint_guardless(symbool: torch.SymBool) -> torch.SymInt:
...
+def _is_derived_dim(dim):
+    return dim.__class__.__name__ == "_DerivedDim"
+
+def _is_dim(dim):
+    return dim.__class__.__name__ == "_Dim"
+
+def _solve_for_root(expr, val):
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:1344]: could you use `try_solve` instead?

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1879,9 +1889,79 @@ def remove_redundant_dynamic_results(self):
...
+            )
+
+        # collect roots
+        modified_roots: Set[str] = {
+            name_to_dim[k].root.__name__
+            for k in results
+            if k in name_to_dim and _is_derived_dim(name_to_dim[k])
+        }
+
+        # evaluate each root for either concrete val or min/max range
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:1919]: The algo here is hard to follow, please add a running example. It's intricate logic, having the transformation explicitly laid out in comments, and how it's achieved with every major step of the transformation, would be good.

Also maybe add comments on why just suggesting fixes to the roots is sufficient.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -1919,6 +2001,18 @@ def relation_with_digit(expr, op, digit):
                     assert op == "=="
                     results[expr]["eq"] = digit
 
+            # retrieve dynamic shapes
+            name_to_dim = {}
+            for dim in pytree.tree_flatten(
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:2006]: Isn't there a different way to collect all dims? Like, have you checked the contents of `source_name_to_debug_name`?

Comment (pianpwk) [reply torch/fx/experimental/symbolic_shapes.py:2006]: I think the main issue here is collecting root dims which aren't attached to a size

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -1883,9 +1883,192 @@ def remove_redundant_dynamic_results(self):
...
+        # {"dx": {"eq": 3*_dx+1, "min": 4, "max": 10}, "dy": dx+1, "dz": dx+2}
+        # we want instead:
+        # {"_dx": {"min": 1, "max": 4}, "dx": 3*_dx+1, "dy": 3*_dx+2, "dz": 3*_dx+3}
+        introduced_roots: Dict[str, str] = {}  # map new root -> old root
+        for k, c in list(results.items()):
+            if "eq" in c and isinstance(c["eq"], sympy.Expr):  # derived dim
+                root = list(c["eq"].free_symbols)[0]
+                if not str(root) in name_to_dim:
+                    introduced_roots[str(root)] = k
+                    # calculate necessary min & max
```
Comment (pianpwk) [inline torch/fx/experimental/symbolic_shapes.py:1968]: moved this code from prettify_results, existed for newly introduced roots

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -1976,21 +2176,11 @@ def relation_with_digit(expr, op, digit):
...
             }
             for k, c in results.items():
-                # if k not in debug_names:
-                #     continue
                 if "eq" in c:
                     other = c["eq"]
                     if isinstance(other, int):
                         others.append(f"{k} = {other}")
                     elif self._is_supported_equivalence(other):
-                        s = next(iter(other.free_symbols))
```
Comment (pianpwk) [inline torch/fx/experimental/symbolic_shapes.py:1986]: this handles newly introduced roots, move to new function

**Review by avikchaudhuri — APPROVED** [review]
Trusting the tests and the summary, the changes are great! Mostly skimmed over the code.

**Comment by pianpwk** [comment]
> Trusting the tests and the summary, the changes are great! Mostly skimmed over the code.

Thank you!
